aws/src/main/java/org/apache/iceberg/aws/s3/signer/S3ObjectMapper.java

```diff
@@ -42,26 +43,24 @@
 
 public class S3ObjectMapper {
 
-  private static final JsonFactory FACTORY = new JsonFactory();
-  private static final ObjectMapper MAPPER = new ObjectMapper(FACTORY);
-  private static volatile boolean isInitialized = false;
+  private static final AtomicReference<ObjectMapper> REFERENCE = new AtomicReference<>();
 
   private S3ObjectMapper() {}
 
   static ObjectMapper mapper() {
-    if (!isInitialized) {
-      synchronized (S3ObjectMapper.class) {
-        if (!isInitialized) {
-          MAPPER.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);
-          MAPPER.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
-          MAPPER.setPropertyNamingStrategy(PropertyNamingStrategies.KebabCaseStrategy.INSTANCE);
-          MAPPER.registerModule(initModule());
-          isInitialized = true;
-        }
-      }
+    if (null == REFERENCE.get()) {
```

Comment:
Thanks, the change now looks correct.

I'm a bit on the fence about it, as now different threads init'ing at the same time will start initializing multiple mappers, which allocates some memory for nothing, but before they would just wait. 

Curious is there some added advantage of atomicReference over double-check locking, other than code correctness?

Reply:
Yeah it's a good question. It really is more of a cleanliness thing imo, it just felt a bit unnecessary to have double check locking when atomic reference already encapsulates a few of the aspects we care about. 

For avoiding the unnecessary work being done by one thread in the situation you described, could we make the mapper() method itself synchronized? That way two threads cannot init at the same time anyways, and the second thread will see the atomic reference is set and not do any work.

Reply:
I also have doubt that the new approach added advantages over the double checked locking.

Reply:
> Curious is there some added advantage of atomicReference over double-check locking, other than code cleanliness?

there's no obvious advantage being added, other than slightly less code and the fact that you don't need to think about `volatile` variables (as double-checked locking only works with `volatile`)

Reply:
given that we use double-locking everywhere else in the codebase and this doesn't provide a clear improvement, I'm going to close it.